packages/template-retail-react-app/app/pages/product-list/index.jsx

```diff
@@ -178,6 +181,30 @@ const ProductList = (props) => {
         }
     }
 
+    /**************** Einstein ****************/
+    useEffect(() => {
+        if (searchQuery && productSearchResult && productSearchResult.hits) {
+            const products = productSearchResult.hits.map((productSearchItem) => {
+                const {productId, sku = '', altId = '', altIdType = ''} = productSearchItem
+                return {
+                    id: productId,
+                    sku,
+                    altId,
+                    altIdType
+                }
+            })
+            einstein.sendViewSearch(searchQuery, products)
+        } else if (category && productSearchResult && productSearchResult.hits) {
+            // Categories update asynchronously from category search results. To ensure we are
+            // sending a viewCategory only after both category and search results update, we track
+            // the previous category's search results in a useRef.
+            if (previousCategoryProducts.current != productSearchResult.hits) {
+                einstein.sendViewCategory(category, productSearchResult)
+                previousCategoryProducts.current = productSearchResult.hits
+            }
+        }
+    }, [searchQuery, category, productSearchResult])
+
```

Comment:
In general in a single page app it's difficult to really pinpoint what a page view is. Since we are loading the page component once, then potentially looking at other categories or doing other searches that don't necessarily cause a component re-mount that we can tie into the trigger these einstein events.

I think there are opportunities to make this logic a little simpler. Reach out to me and we can go through some ideas I have that are more easily communicated through a chat.

Reply:
Thanks for the sync, @bendvc.

The logic around this part is simplified now.